keps/sig-api-machinery/3610-env-injection/README.md

```diff
@@ -0,0 +1,855 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This KEP defines a new annotation `glabel.env` and introduces a mechanism that can override current environment variables on pods.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+This seems to be a common requirement.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- inject env to all containers in namespace
+- skip the env if container already has set
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+Add a new Admission Controller: GlobalEnv. Users can enable or disable it on the apiserver side.
+If enabled, the admission controller will add env according to namespace annotation to every pod's containers in the namespace wide.
+A namespace annotation named `global.env`, works for pods in the namespace.
+
+### Other Proposals and Concerns
+
+**annotation vs configmap**
+Annotation is simpler, So I choose to use an annotation here.
+
+- For annotation, we should divide the string by `,` and get the key and value by `=`, for instance, "key1=value1,key2=value2".
```

Comment:
We have an admission controller added to the intree api server on the kubernetes that we manage in house and this format of `,` and `=` for separation got too complicated too soon with all the patterns and ability to patch them on the go easily. Instead we decided to switch to using `env-to-inject/<env-name>: <env-value>` for the annotation. And the values were `base64` encoded always to be able to inject complicated patterns.

Reply:
It seems better if we choose to use annotation.
Or we can use configmap for complicated cases.

Reply:
I would love to see the annotation over the config map as there will be a pre-built assumption that it will have a specific name. Not that it is a bad thing. But, that is just my feeling,